Sources/TSCBasic/Thread.swift

```diff
@@ -85,3 +95,60 @@ final private class ThreadImpl: Foundation.Thread {
 // Thread on Linux supports closure so just use it directly.
 typealias ThreadImpl = Foundation.Thread
 #endif
+
+protocol Defaultable {
+    static var defaultValue: Self { get }
+}
+
+extension Optional: Defaultable {
+    static var defaultValue: Optional<Wrapped> { .none }
+}
+
+/// `ThreadLocal` properties are  thread-specific. Every thread has its own instance of the wrapped property.
+@propertyWrapper final class ThreadLocal<Value: Defaultable> {
```

Comment:
ThreadLocal as a property wrapper seems very useful, but feels too general for this PR.  It feels more like a language-level feature, and we can discuss whether that should be in ToolsSupportCore.

For this case, with the file locking, would it be possible to use Foundation.Thread's `threadDictionary` directly?

Reply:
+1 for proposing such generalizations as separate followup PR so can be discussed individually

Reply:
I think I would be easier to leave `FileLock` thread-unsafe for the moment and to fix this issue once `@ThreadLocal` and `@AutoClosing` are available in SwiftSystem or ToolSupportCore.